emailproxy.py

```diff
@@ -279,6 +289,58 @@ def _load():
             AppConfig._GLOBALS = configparser.SectionProxy(AppConfig._PARSER, APP_SHORT_NAME)
         AppConfig._SERVERS = [s for s in config_sections if CONFIG_SERVER_MATCHER.match(s)]
         AppConfig._ACCOUNTS = [s for s in config_sections if '@' in s]
+
+        def fetch_aws_secrets():
+            if 'boto3' not in sys.modules:
+                Log.error('Error: client configuration for some accounts includes an aws_secret'
+                          ' parameter, but the app has been loaded without the --aws-secrets option.')
+                raise ValueError("--aws-secrets must be specified if account configuration includes aws_secret")
```

Comment:
I'm a little reluctant to exit the proxy when errors or misconfigurations occur (both here and other places) unless they are absolutely critical. Perhaps just return here without loading?

Reply:
@simonrob I understand your reluctance! Let's make a conscious decision about the expected behaviour when:

- the config for an account specifies `aws_secret:`, so the user has indicated that they want to store their tokens in AWS Secrets Manager.
- the user has called the proxy without the `--aws-secrets` argument, so `boto3` isn't loaded and the proxy cannot use the AWS Secrets Manager.

I see three potential behaviours:
1. Throw an error explaining the issue and exit.
2. Log a warning explaining the issue and then run the proxy. Cache the tokens in memory, but do not write them to the local config file.
3. Log a warning explaining the issue and then run the proxy. Write the tokens to the local config file (as if the user had not specified `aws_secret:`.

In my personal view: 3 is a bad solution, because it defies a user configuration which says "don't store the tokens in my local config file." Options 1 and 2 both seem reasonable to me.

I would lean toward solution 1 (immediately throw an error when the proxy is launched and exit, as currently implemented) because this issue is only going to be encountered by advanced users who are setting up a custom config after reading the docs. They haven't quite gotten their config/command right, because they are instructing the proxy to use AWS Secrets but haven't passed the argument to do so. I think an error in this case is most helpful, since it gives the user immediate feedback to correct their custom config/command.

Solution 2 meets your objective to keep the proxy running whenever possible. I think it is more likely to result in confusion: there will be cases where a user is inattentive, doesn't notice the warning in the logs, the proxy is working fine, and then hours/days/weeks later they wonder why the tokens aren't stored in AWS Secrets Manager in a scenario where they are forced to reauthenticate. They do some debugging and notice the warning about the misconfiguration.

In sum: in this scenario, a user has misconfigured an advanced setting, and I favour solution 1 because it gives immediate feedback prompting them to fix the configuration. But I'd be happy to switch to whatever solution you prefer!

Reply:
Thanks for the summary of options. I do agree with your reasoning in the first case, but I wonder what the best action would be if the sequence of events is, say:

1: Load the proxy without `--aws-secrets` with a set of accounts that do not have any `aws_secret` values
2: Some time later, edit the configuration file to add `aws_secret` values to a few accounts
3: Reload the configuration file (from the GUI) or send `SIGHUP`

Reply:
@simonrob That's an excellent question – I wasn't aware that the config could be reloaded on the fly until today. I liked the idea of exiting the proxy with an error immediately upon launch, but I like it much less if it happens in scenarios that aren't immediately upon launch.

Given your point, I think the best course of action is Solution 2: log a warning explaining the issue, continue running the proxy, cache the tokens in memory, but do not write them to the local config file. The proxy keeps running, but will require re-authentication when it is next restarted.

I'll go ahead and implement that behaviour in my next revision, and test it along with testing the AWS errors to see if we can get a more precise solution there.